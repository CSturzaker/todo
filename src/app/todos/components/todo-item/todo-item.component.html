<div class="todo">
  <mat-checkbox
    class="todo-done"
    [checked]="todo.done"
    (change)="toggleTodo(todo)"
  ></mat-checkbox>
  <span
    class="todo-label"
    [ngClass]="{ 'todo-label-done': todo.done }"
    (click)="toggleTodo(todo)"
  >
    {{ todo.name }}
  </span>
  <button mat-icon-button (click)="removeTodo(todo)" aria-label="Delete">
    <mat-icon>delete</mat-icon>
  </button>
</div>
