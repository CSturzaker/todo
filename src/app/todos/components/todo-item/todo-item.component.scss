@import '../../../../variables.scss';

.todo {
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 0 15px;
  &:hover {
    background-color: $purple;
  }
  cursor: pointer;

  .todo-done {
    margin: 0 10px 0 0;
  }

  .todo-label {
    position: relative;
    top: 2px;

    &.todo-label-done {
      text-decoration: line-through;
      opacity: 0.5;
    }
  }

  mat-icon {
    color: $dark-purple;
  }

  button {
    margin-left: auto;
  }
}
