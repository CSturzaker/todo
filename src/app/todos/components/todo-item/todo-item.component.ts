import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core'
import { Todo } from '../../models'

@Component({
  selector: 'todo-todo-item',
  templateUrl: './todo-item.component.html',
  styleUrls: ['./todo-item.component.scss'],
})
export class TodoItemComponent implements OnInit {
  @Input()
  todo: Todo
  @Output() toggleTodoEvent = new EventEmitter<Todo>()
  @Output() removeTodoEvent = new EventEmitter<Todo>()

  constructor() {}

  ngOnInit() {}

  toggleTodo(todo: Todo) {
    this.toggleTodoEvent.emit(todo)
  }

  removeTodo(todo: Todo) {
    this.removeTodoEvent.emit(todo)
  }
}
