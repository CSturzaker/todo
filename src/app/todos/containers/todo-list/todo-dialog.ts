import { Component, Inject } from '@angular/core'
import {
  MatDialog,
  MatDialogRef,
  MAT_DIALOG_DATA,
} from '@angular/material/dialog'

@Component({
  selector: 'todo-todo-dialog',
  templateUrl: 'todo-dialog.html',
  styles: [
    `
      mat-form-field {
        width: 100%;
      }
    `,
  ],
})
export class TodoDialogComponent {
  constructor(
    public dialogRef: MatDialogRef<TodoDialogComponent>,
    @Inject(MAT_DIALOG_DATA) public data: object
  ) {}

  onNoClick(): void {
    this.dialogRef.close()
  }
}
