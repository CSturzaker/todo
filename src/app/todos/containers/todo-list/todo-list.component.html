<div class="todo-list-container">
  <todo-todo-item
    *ngFor="let todo of todoItems$ | async"
    [todo]="todo"
    (toggleTodoEvent)="toggleTodo(todo)"
    (removeTodoEvent)="removeTodo(todo)"
  ></todo-todo-item>
</div>
<div class="footer">
  <button mat-icon-button aria-label="Add a to-do">
    <mat-icon>add_circle</mat-icon>
  </button>
  Add a to-do
</div>
