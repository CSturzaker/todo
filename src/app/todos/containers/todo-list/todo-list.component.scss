@import '../../../../variables.scss';

.todo-list-container {
  background-color: $white;
  max-height: 300px;
}

.footer {
  color: $dark-purple;
  display: flex;
  flex-direction: row;
  align-items: center;
  cursor: pointer;
}
