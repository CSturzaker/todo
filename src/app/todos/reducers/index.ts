import {
  createSelector,
  createFeatureSelector,
  Action,
  combineReducers,
} from '@ngrx/store'
import * as fromRoot from '../../state'
import * as fromTodos from '../reducers/todo.reducer'

export const todoFeatureKey = 'todo'

export interface TodoState {
  [fromTodos.todoFeatureKey]: fromTodos.State
}

export interface State extends fromRoot.State {
  [todoFeatureKey]: TodoState
}

export function reducers(state: TodoState | undefined, action: Action) {
  return combineReducers({
    [fromTodos.todoFeatureKey]: fromTodos.reducer,
  })(state, action)
}

export const selectTodoState = createFeatureSelector<State, TodoState>(
  todoFeatureKey
)

export const selectTodoStatusState = createSelector(
  selectTodoState,
  (state: TodoState) => state.todo
)

export const selectTodos = createSelector(
  selectTodoStatusState,
  fromTodos.getTodoItems
)
