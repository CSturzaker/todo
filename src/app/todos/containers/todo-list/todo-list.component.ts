import { Component, ChangeDetectionStrategy } from '@angular/core'
import { Store, select } from '@ngrx/store'

import { Observable } from 'rxjs'

import * as TodoActions from '../../actions/todo.actions'

import { Todo } from '../../models'

import * as fromTodos from '../../reducers'

@Component({
  selector: 'todo-todo-list',
  templateUrl: './todo-list.component.html',
  styleUrls: ['./todo-list.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class TodoListComponent {
  todoItems$: Observable<Todo[]> = this.store.pipe(
    select(fromTodos.selectTodos)
  )
  newTodo = ''

  constructor(private store: Store<fromTodos.State>) {}

  addTodo() {
    this.store.dispatch(TodoActions.addTodo(this.newTodo))
    this.newTodo = ''
  }

  toggleTodo(todo: Todo) {
    const { id } = todo
    this.store.dispatch(TodoActions.toggleTodo({ id }))
  }

  removeTodo(todo: Todo) {
    const { id } = todo
    this.store.dispatch(TodoActions.removeTodo({ id }))
  }
}
