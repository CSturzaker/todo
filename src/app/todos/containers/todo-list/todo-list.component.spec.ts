import { v4 as uuid } from 'uuid'
import { ComponentFixture, TestBed } from '@angular/core/testing'
import { provideMockStore, MockStore } from '@ngrx/store/testing'
import { Store } from '@ngrx/store'
import {
  MatDialogModule,
  MatCheckboxModule,
  MatIconModule,
} from '@angular/material'
import { MatDialog } from '@angular/material/dialog'

import * as fromTodos from '../../reducers'

import { TodoListComponent } from './todo-list.component'
import { TodoItemComponent } from '../../components/todo-item/todo-item.component'

describe('TodoListComponent', () => {
  let component: TodoListComponent
  let fixture: ComponentFixture<TodoListComponent>
  let store: MockStore<fromTodos.State>

  const todoItems = [
    { id: uuid(), name: 'Shortlist features for MVP', done: false },
    { id: uuid(), name: 'Launch PPC campaign with new creative', done: false },
    {
      id: uuid(),
      name: 'Define audience breakdown with new data',
      done: false,
    },
    { id: uuid(), name: 'Launch demo page for SEO analysis', done: true },
  ]

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [MatDialogModule, MatCheckboxModule, MatIconModule],
      declarations: [TodoListComponent, TodoItemComponent],
      providers: [
        provideMockStore({
          selectors: [{ selector: fromTodos.selectTodos, value: todoItems }],
        }),
        { provide: MatDialog, useValue: {} },
      ],
    })

    fixture = TestBed.createComponent(TodoListComponent)
    component = fixture.componentInstance

    store = TestBed.get(Store)

    spyOn(store, 'dispatch')
  })

  it('should create', () => {
    fixture.detectChanges()

    expect(fixture).toMatchSnapshot()
  })
})
