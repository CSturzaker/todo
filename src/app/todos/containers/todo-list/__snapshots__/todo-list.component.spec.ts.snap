// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListComponent should create 1`] = `
<todo-todo-list
  dialog={[Function Object]}
  newTodo=""
  store={[Function MockStore]}
  todoItems$={[Function Store]}
>
  <div
    class="todo-list-container"
  >
    
    <todo-todo-item
      ng-reflect-todo="[object Object]"
    >
      <div
        class="todo"
      >
        <mat-checkbox
          class="todo-done mat-checkbox mat-accent"
          id="mat-checkbox-1"
          ng-reflect-checked="false"
        >
          <label
            class="mat-checkbox-layout"
            for="mat-checkbox-1-input"
          >
            <div
              class="mat-checkbox-inner-container mat-checkbox-inner-container-no-side-margin"
            >
              <input
                aria-checked="false"
                class="mat-checkbox-input cdk-visually-hidden"
                id="mat-checkbox-1-input"
                tabindex="0"
                type="checkbox"
              />
              <div
                class="mat-checkbox-ripple mat-ripple"
                matripple=""
                ng-reflect-animation="[object Object]"
                ng-reflect-centered="true"
                ng-reflect-disabled="false"
                ng-reflect-radius="20"
                ng-reflect-trigger="[object HTMLLabelElement]"
              >
                <div
                  class="mat-ripple-element mat-checkbox-persistent-ripple"
                />
              </div>
              <div
                class="mat-checkbox-frame"
              />
              <div
                class="mat-checkbox-background"
              >
                <svg
                  class="mat-checkbox-checkmark"
                  focusable="false"
                  version="1.1"
                  viewBox="0 0 24 24"
                  xml:space="preserve"
                >
                  <path
                    class="mat-checkbox-checkmark-path"
                    d="M4.1,12.7 9,17.6 20.3,6.3"
                    fill="none"
                    stroke="white"
                  />
                </svg>
                <div
                  class="mat-checkbox-mixedmark"
                />
              </div>
            </div>
            <span
              class="mat-checkbox-label"
            >
              <span
                style="display:none"
              >
                 
              </span>
            </span>
          </label>
        </mat-checkbox>
        <span
          class="todo-label"
          ng-reflect-klass="todo-label"
          ng-reflect-ng-class="[object Object]"
        >
           Shortlist features for MVP 
        </span>
        <button
          aria-label="Delete"
          mat-icon-button=""
        >
          <mat-icon
            aria-hidden="true"
            class="mat-icon notranslate material-icons mat-icon-no-color"
            role="img"
          >
            delete
          </mat-icon>
        </button>
      </div>
    </todo-todo-item>
    <todo-todo-item
      ng-reflect-todo="[object Object]"
    >
      <div
        class="todo"
      >
        <mat-checkbox
          class="todo-done mat-checkbox mat-accent"
          id="mat-checkbox-2"
          ng-reflect-checked="false"
        >
          <label
            class="mat-checkbox-layout"
            for="mat-checkbox-2-input"
          >
            <div
              class="mat-checkbox-inner-container mat-checkbox-inner-container-no-side-margin"
            >
              <input
                aria-checked="false"
                class="mat-checkbox-input cdk-visually-hidden"
                id="mat-checkbox-2-input"
                tabindex="0"
                type="checkbox"
              />
              <div
                class="mat-checkbox-ripple mat-ripple"
                matripple=""
                ng-reflect-animation="[object Object]"
                ng-reflect-centered="true"
                ng-reflect-disabled="false"
                ng-reflect-radius="20"
                ng-reflect-trigger="[object HTMLLabelElement]"
              >
                <div
                  class="mat-ripple-element mat-checkbox-persistent-ripple"
                />
              </div>
              <div
                class="mat-checkbox-frame"
              />
              <div
                class="mat-checkbox-background"
              >
                <svg
                  class="mat-checkbox-checkmark"
                  focusable="false"
                  version="1.1"
                  viewBox="0 0 24 24"
                  xml:space="preserve"
                >
                  <path
                    class="mat-checkbox-checkmark-path"
                    d="M4.1,12.7 9,17.6 20.3,6.3"
                    fill="none"
                    stroke="white"
                  />
                </svg>
                <div
                  class="mat-checkbox-mixedmark"
                />
              </div>
            </div>
            <span
              class="mat-checkbox-label"
            >
              <span
                style="display:none"
              >
                 
              </span>
            </span>
          </label>
        </mat-checkbox>
        <span
          class="todo-label"
          ng-reflect-klass="todo-label"
          ng-reflect-ng-class="[object Object]"
        >
           Launch PPC campaign with new creative 
        </span>
        <button
          aria-label="Delete"
          mat-icon-button=""
        >
          <mat-icon
            aria-hidden="true"
            class="mat-icon notranslate material-icons mat-icon-no-color"
            role="img"
          >
            delete
          </mat-icon>
        </button>
      </div>
    </todo-todo-item>
    <todo-todo-item
      ng-reflect-todo="[object Object]"
    >
      <div
        class="todo"
      >
        <mat-checkbox
          class="todo-done mat-checkbox mat-accent"
          id="mat-checkbox-3"
          ng-reflect-checked="false"
        >
          <label
            class="mat-checkbox-layout"
            for="mat-checkbox-3-input"
          >
            <div
              class="mat-checkbox-inner-container mat-checkbox-inner-container-no-side-margin"
            >
              <input
                aria-checked="false"
                class="mat-checkbox-input cdk-visually-hidden"
                id="mat-checkbox-3-input"
                tabindex="0"
                type="checkbox"
              />
              <div
                class="mat-checkbox-ripple mat-ripple"
                matripple=""
                ng-reflect-animation="[object Object]"
                ng-reflect-centered="true"
                ng-reflect-disabled="false"
                ng-reflect-radius="20"
                ng-reflect-trigger="[object HTMLLabelElement]"
              >
                <div
                  class="mat-ripple-element mat-checkbox-persistent-ripple"
                />
              </div>
              <div
                class="mat-checkbox-frame"
              />
              <div
                class="mat-checkbox-background"
              >
                <svg
                  class="mat-checkbox-checkmark"
                  focusable="false"
                  version="1.1"
                  viewBox="0 0 24 24"
                  xml:space="preserve"
                >
                  <path
                    class="mat-checkbox-checkmark-path"
                    d="M4.1,12.7 9,17.6 20.3,6.3"
                    fill="none"
                    stroke="white"
                  />
                </svg>
                <div
                  class="mat-checkbox-mixedmark"
                />
              </div>
            </div>
            <span
              class="mat-checkbox-label"
            >
              <span
                style="display:none"
              >
                 
              </span>
            </span>
          </label>
        </mat-checkbox>
        <span
          class="todo-label"
          ng-reflect-klass="todo-label"
          ng-reflect-ng-class="[object Object]"
        >
           Define audience breakdown with new data 
        </span>
        <button
          aria-label="Delete"
          mat-icon-button=""
        >
          <mat-icon
            aria-hidden="true"
            class="mat-icon notranslate material-icons mat-icon-no-color"
            role="img"
          >
            delete
          </mat-icon>
        </button>
      </div>
    </todo-todo-item>
    <todo-todo-item
      ng-reflect-todo="[object Object]"
    >
      <div
        class="todo"
      >
        <mat-checkbox
          class="todo-done mat-checkbox mat-accent mat-checkbox-checked"
          id="mat-checkbox-4"
          ng-reflect-checked="true"
        >
          <label
            class="mat-checkbox-layout"
            for="mat-checkbox-4-input"
          >
            <div
              class="mat-checkbox-inner-container mat-checkbox-inner-container-no-side-margin"
            >
              <input
                aria-checked="true"
                class="mat-checkbox-input cdk-visually-hidden"
                id="mat-checkbox-4-input"
                tabindex="0"
                type="checkbox"
              />
              <div
                class="mat-checkbox-ripple mat-ripple"
                matripple=""
                ng-reflect-animation="[object Object]"
                ng-reflect-centered="true"
                ng-reflect-disabled="false"
                ng-reflect-radius="20"
                ng-reflect-trigger="[object HTMLLabelElement]"
              >
                <div
                  class="mat-ripple-element mat-checkbox-persistent-ripple"
                />
              </div>
              <div
                class="mat-checkbox-frame"
              />
              <div
                class="mat-checkbox-background"
              >
                <svg
                  class="mat-checkbox-checkmark"
                  focusable="false"
                  version="1.1"
                  viewBox="0 0 24 24"
                  xml:space="preserve"
                >
                  <path
                    class="mat-checkbox-checkmark-path"
                    d="M4.1,12.7 9,17.6 20.3,6.3"
                    fill="none"
                    stroke="white"
                  />
                </svg>
                <div
                  class="mat-checkbox-mixedmark"
                />
              </div>
            </div>
            <span
              class="mat-checkbox-label"
            >
              <span
                style="display:none"
              >
                 
              </span>
            </span>
          </label>
        </mat-checkbox>
        <span
          class="todo-label todo-label-done"
          ng-reflect-klass="todo-label"
          ng-reflect-ng-class="[object Object]"
        >
           Launch demo page for SEO analysis 
        </span>
        <button
          aria-label="Delete"
          mat-icon-button=""
        >
          <mat-icon
            aria-hidden="true"
            class="mat-icon notranslate material-icons mat-icon-no-color"
            role="img"
          >
            delete
          </mat-icon>
        </button>
      </div>
    </todo-todo-item>
  </div><div
    class="footer"
  >
    <button
      aria-label="Add a to-do"
      mat-icon-button=""
    >
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons mat-icon-no-color"
        role="img"
      >
        add_circle
      </mat-icon>
    </button>
     Add a to-do
  
  </div>
</todo-todo-list>
`;
